src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,468 @@
+  return res;
+}
+
+static GVariant *
+machines_get_property (GDBusConnection *connection,
+                       const gchar *sender,
+                       const gchar *object_path,
+                       const gchar *interface_name,
+                       const gchar *property_name,
+                       GError **error,
+                       gpointer user_data)
+{
+  g_return_val_if_fail (property_name != NULL, NULL);
+
+  if (g_str_equal (property_name, "Machines"))
+    return get_machines ();
+  else
+    g_return_val_if_reached (NULL);
+}
+
+static void
+machines_method_call (GDBusConnection *connection,
+                      const gchar *sender,
+                      const gchar *object_path,
+                      const gchar *interface_name,
+                      const gchar *method_name,
+                      GVariant *parameters,
+                      GDBusMethodInvocation *invocation,
+                      gpointer user_data)
+{
+  if (g_str_equal (method_name, "Update"))
+    {
+      const gchar *filename, *hostname;
+      GVariant * info_v;
+      JsonNode *info_json = NULL;
+      GError *error = NULL;
+
+      g_variant_get (parameters, "(&s&s@a{sv})", &filename, &hostname, &info_v);
+      info_json = json_gvariant_serialize (info_v);
+      g_debug ("Updating %s for machine %s", filename, hostname);
+      g_variant_unref (info_v);
+
+      if (update_machine (filename, hostname, info_json, &error))
+        g_dbus_method_invocation_return_value (invocation, NULL);
+      else
+        g_dbus_method_invocation_take_error (invocation, error);
+      json_node_free (info_json);
+    }
+  else
+    g_return_if_reached ();
+}
+
+/**
+ * notify_properties:
+ * @user_data: GDBusConnection to which updates get sent
+ *
+ * Send a PropertiesChanged signal for invalidating the Machines property. This
+ * avoids parsing files and constructing the property when nobody is listening.
+ * This gets called in reaction to changed *.json configuration files.
+ */
+static gboolean
+notify_properties (gpointer user_data)
+{
+  GDBusConnection *connection = user_data;
+  GVariant *signal_value;
+  GVariantBuilder builder;
+  GError *error = NULL;
+
+  /* reset pending counter before we do any actual work, to avoid races */
+  pending_updates = 0;
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("as"));
+  g_variant_builder_add (&builder, "s", "Machines");
+  signal_value = g_variant_ref_sink (g_variant_new ("(sa{sv}as)", "cockpit.Machines", NULL, &builder));
+
+  g_dbus_connection_emit_signal (connection,
+                                 NULL,
+                                 "/machines",
+                                 "org.freedesktop.DBus.Properties",
+                                 "PropertiesChanged",
+                                 signal_value,
+                                 &error);
+  if (error != NULL)
+    {
+      g_critical ("failed to send PropertiesChanged signal: %s", error->message);
```

Comment:
This can happen during shutdown when the DBus connection is closed. I don't think it's a g_critical() unless we account for this:

```
src/bridge/cockpitdbuscache.c:      g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CLOSED))
```